### PR TITLE
fixed unresponsive server  #3283

### PR DESCRIPTION
**Pull request by Krishanu230**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Eomm** [comment]
Could you add a test for this bugfix?

File: lib/reply.js [hunk 1]
```diff
@@ -377,8 +377,8 @@ function preserializeHookEnd (err, request, reply, payload) {
 }
 
 function onSendHook (reply, payload) {
-  reply[kReplySent] = true
   if (reply.context.onSend !== null) {
+    reply[kReplySent] = true
```
Comment (climba03003) [inline lib/reply.js:381]: I think it should be completely removed as this value will be set inside `onSendHookRunner` and `onSendEnd`

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you please add a unit test?

File: test/reply-error.test.js [hunk 2]
```diff
@@ -525,3 +525,73 @@ test('should trigger error handlers if a sync route throws any non-error object'
   t.equal(reply.statusCode, 500)
   t.equal(JSON.parse(reply.body).foo, 'bar')
 })
+
+// Issue 3283 https://github.com/fastify/fastify/issues/3283
```
Comment (Eomm) [inline test/reply-error.test.js:529]: This can be found looking the git history/blame
```suggestion
```

File: test/reply-error.test.js [hunk 3]
```diff
@@ -525,3 +525,73 @@ test('should trigger error handlers if a sync route throws any non-error object'
...
+      {
+        error: 'Internal Server Error',
+        message: 'Attempted to send payload of invalid type \'object\'. Expected a string or Buffer.',
+        statusCode: 500,
+        code: 'FST_ERR_REP_INVALID_PAYLOAD_TYPE'
+      },
+      JSON.parse(res.payload)
+    )
+  })
+})
```
Comment (Eomm) [inline test/reply-error.test.js:577]: I changed a bit this test to see more clearly that the server hangs out without your fix 👍🏽 

```suggestion
test('setting content-type on reply object should not hang the server case 2', async t => {
  t.plan(1)
  const fastify = Fastify()

  fastify.get('/', (req, reply) => {
    reply
      .code(200)
      .headers({ 'content-type': 'text/plain; charset=utf-8' })
      .send({ bar: 'foo', baz: 'foobar' })
  })

  try {
    await fastify.listen(0)
    const res = await fastify.inject({
      url: '/',
      method: 'GET'
    })
    t.same({
      error: 'Internal Server Error',
      message: 'Attempted to send payload of invalid type \'object\'. Expected a string or Buffer.',
      statusCode: 500,
      code: 'FST_ERR_REP_INVALID_PAYLOAD_TYPE'
    },
    res.json())
  } catch (error) {
    t.error(error)
  } finally {
    await fastify.close()
  }
})
```

**Review by Eomm — APPROVED** [review]

File: lib/reply.js (hunk 1 above)
Comment (Krishanu230) [reply lib/reply.js:381]: If I remove it completely, The following tests fail:
1. test/hooks-async.test.js 5 failed
2. test/hooks.test.js 1 failed

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**